tianshou/data/collector.py

```diff
@@ -274,110 +242,70 @@ def collect(self,
-                    self._cached_buf[i].add(**self.data[i])
-                    if self.data.done[i]:
-                        if n_step != 0 or np.isscalar(n_episode) or \
-                                cur_episode[i] < n_episode[i]:
-                            cur_episode[i] += 1
-                            reward_sum += self.reward[i]
-                            length_sum += self.length[i]
-                            if self._cached_buf:
-                                cur_step += len(self._cached_buf[i])
-                                if self.buffer is not None:
-                                    self.buffer.update(self._cached_buf[i])
-                        self.reward[i], self.length[i] = 0., 0
-                        if self._cached_buf:
-                            self._cached_buf[i].reset()
-                        self._reset_state(i)
-                obs_next = self.data.obs_next
-                if sum(self.data.done):
-                    env_ind = np.where(self.data.done)[0]
-                    obs_reset = self.env.reset(env_ind)
-                    if self.preprocess_fn:
-                        obs_next[env_ind] = self.preprocess_fn(
-                            obs=obs_reset).get('obs', obs_reset)
-                    else:
-                        obs_next[env_ind] = obs_reset
-                self.data.obs_next = obs_next
-                if n_episode != 0:
-                    if isinstance(n_episode, list) and \
-                            (cur_episode >= np.array(n_episode)).all() or \
-                            np.isscalar(n_episode) and \
-                            cur_episode.sum() >= n_episode:
-                        break
-            else:  # single buffer, without cache_buffer
-                if self.buffer is not None:
-                    self.buffer.add(**self.data[0])
-                cur_step += 1
-                if self.data.done[0]:
-                    cur_episode += 1
-                    reward_sum += self.reward[0]
-                    length_sum += self.length[0]
-                    self.reward, self.length = 0., np.zeros(self.env_num)
-                    self.data.state = Batch()
-                    obs_next = self._make_batch(self.env.reset())
-                    if self.preprocess_fn:
-                        obs_next = self.preprocess_fn(obs=obs_next).get(
-                            'obs', obs_next)
-                    self.data.obs_next = obs_next
-                if n_episode != 0 and cur_episode >= n_episode:
+            for i in range(self.env_num):
+                self._cached_buf[i].add(**self.data[i])
+                if self.data.done[i]:
+                    if n_step or np.isscalar(n_episode) or \
+                            episode_count[i] < n_episode[i]:
+                        episode_count[i] += 1
+                        reward_total += np.asarray(
+                            self._cached_buf[i].rew).sum(axis=0)
```

Comment:
What is the type of `self._cached_buf[i].rew` ? Batch already support `np.sum`. To me it can only be a Batch or `np.ndarray`, so using `asarray` is inecessary.

Reply:
It is a list, because the cache buffer is based on list.

Reply:
It bypasses the setattr method

Reply:
Ah yes it remember now.